PSReadLine/PlatformWindows.cs

```diff
@@ -1015,4 +1015,34 @@ private static void TerminateStragglers()
             }
         }
     }
+
+    /// <remarks>
+    /// This method helps to find the active keyboard layout in a terminal process that controls the current console
+    /// application. For now, we assume that at the moment when we are asked to process a keyboard shortcut, the
+    /// process owning the foreground window has to be the terminal process controlling the current console.
+    /// </remarks>
+    public static IntPtr GetConsoleKeyboardLayout()
+    {
+        IntPtr foregroundWindow = GetForegroundWindow();
+        if (foregroundWindow != IntPtr.Zero)
+        {
+            uint tid = GetWindowThreadProcessId(foregroundWindow, out _);
+            if (tid != 0)
+            {
+                return GetKeyboardLayout(tid);
+            }
+        }
+
+        // Fall back to the default keyboard when we failed to find the parent terminal process.
+        return GetKeyboardLayout(0);
```

Comment:
I just took a look at #3078 and realized that we actually circled back ...

There are 2 concerns expressed in that PR (https://github.com/PowerShell/PSReadLine/pull/3078#discussion_r761562034):
- there could be a race condition -- a window popping up by a background task causes `GetForegroundWindow()` to return the wrong handle.
- calling `GetForegroundWindow()` and `GetWindowThreadProcessId` on every keystroke that produces a `ConsoleKeyInfo key` with a control character of `key.KeyChar` is unnecessary overhead, so caching makes sense.

I think we can solve them both by combining the use of `GetForegroundWindow()` and looking up the process tree:
1. I have verified that when running in conhost, the process id returned by `GetWindowThreadProcessId(GetForegroundWindow(), out uint pid)` points to the console application that was first started with the conhost window. Namely, if you start with `cmd -> pwsh`, `pid` points to the `cmd.exe` process; if you start with `pwsh` directly in conhost, then `pid` points to the `pwsh` process.

1. When running in Windows Terminal and VSCode, the `pid` return from above call points to `WindowsTerminal` and `Code` respectively.

1. Note that, all the process `pid` points to in the above scenarios is either the current `pwsh` process or a parent process of the current `pwsh` process. You should be able to find it when walking up the process tree.
   - So, to solve the "_race condition_" concern, we should call `GetWindowThreadProcessId(GetForegroundWindow(), out uint pid)` and then walking up the process tree to verify if we can find `pid` in the tree:
      1. if we can find it, then we are good and should cache the returned thread id to use for the rest of session;
      1. if we cannot find it, then it was a race condition and we simply try again (call `GetWindowThreadProcessId(GetForegroundWindow, ..)` and verify it) until we are good.

   - To solve the "_caching thread id_" concern, all the above "call and verify" operations should be done in the one-time initializer of PSReadLine. Once we find the correct thread id, we will cache it and use it for the rest of the session.

@ForNeVeR and @lzybkr Please let me know what you think.

Reply:
Ok, this suggestion makes sense. I'll try to piece it all together.